// overmind-devtools/src/components/ActionsList/elements.tsx
import styled from '../../styled-components'

export const List = styled.div`
  border-right: 1px solid
    ${({ theme }) => theme.color.fade(theme.color.black, 0.9)};
`

export const ActionItem = styled.div`
  position: relative;
  padding: ${({ theme }) => theme.padding.smallerer};
  cursor: pointer;
  text-indent: 15px;
  border-bottom: 1px solid
    ${({ theme }) => theme.color.fade(theme.color.black, 0.9)};
  :hover {
    background-color: ${({ theme }) =>
      theme.color.fade(theme.color.black, 0.95)};
  }
`

export const ActionColor = styled<
  {
    color: string
  },
  'span'
>('span')`
  position: absolute;
  top: 0;
  right: 0;
  width: 10px;
  height: 100%;
  background-color: ${({ color }) => color};
`

export const ActionSubItem = styled.div`
  padding: ${({ theme }) => theme.padding.smallerer};
  cursor: pointer;
  text-indent: 15px;
  border-bottom: 1px solid
    ${({ theme }) => theme.color.fade(theme.color.black, 0.9)};
  :hover {
    background-color: ${({ theme }) =>
      theme.color.fade(theme.color.black, 0.95)};
  }
`
