// action-chain/src/ActionBase.ts
import { ActionChain, Execution, ExecutionContext } from './ActionChain'

export class StopExecution {
  value: any
  constructor(value) {
    this.value = value
  }
}
export class ActionBase<Context, InitialValue, Value = InitialValue> {
  static nextActionId: number = 0
  private initialActionId: number
  private actionChain: ActionChain<Context>
  private runOperators: (
    value: any,
    executionContext: {
      __execution: Execution
      __path: string[]
    },
    newPath?: string
  ) => any | Promise<any>
  constructor(
    actionChain: ActionChain<Context>,
    initialActionId: number = ActionBase.nextActionId++,
    runOperators?: (
      value: any,
      executionContext: {
        __execution: Execution
        __path: string[]
      },
      newPath?: string
    ) => any | Promise<any>
  ) {
    this.actionChain = actionChain
    this.runOperators = runOperators
    this.initialActionId = initialActionId

    let currentExecutionId = 0
    return Object.assign(function(value) {
      const initialOperator = typeof arguments[1] === 'undefined'
      const newPath = typeof arguments[2] === 'undefined' ? null : arguments[2]
      const executionContext: ExecutionContext = initialOperator
        ? {
            __execution: {
              operatorId: -1,
              actionId: initialActionId,
              executionId: currentExecutionId++,
            },
            __path: [],
          }
        : arguments[1]
      if (initialOperator) {
        actionChain.emit('action:start', {
          actionId: executionContext.__execution.actionId,
          executionId: executionContext.__execution.executionId,
        })
      }
      const returnValue = runOperators
        ? runOperators(value, executionContext, newPath)
        : value

      if (initialOperator && returnValue instanceof Promise) {
        returnValue.then(() => {
          actionChain.emit('action:end', {
            actionId: executionContext.__execution.actionId,
            executionId: executionContext.__execution.executionId,
          })
        })
      } else if (initialOperator) {
        actionChain.emit('action:end', {
          actionId: executionContext.__execution.actionId,
          executionId: executionContext.__execution.executionId,
        })
      }

      return returnValue
    }, this) as any
  }
  getActionChain = () => {
    return this.actionChain
  }
  createOperatorResult = (
    type: string,
    name: string,
    cb: any
  ): [ActionChain<Context>, number, any] => {
    return [
      this.actionChain,
      this.initialActionId,
      (props, executionContext, newPath) => {
        const executionContextWithPath = {
          __execution: executionContext.__execution,
          __path: newPath
            ? executionContext.__path.concat(newPath)
            : executionContext.__path.slice(),
        }
        const path = executionContextWithPath.__path
        const prevResult = this.runOperators
          ? this.runOperators(props, executionContextWithPath)
          : props

        const operatorId = ++executionContextWithPath.__execution.operatorId

        const produceResult = (currentValue) => {
          if (currentValue instanceof StopExecution) {
            return currentValue
          }

          const context = this.actionChain.getContext(executionContextWithPath)

          this.actionChain.emit('operator:start', {
            type,
            name,
            path,
            ...context.__execution,
            operatorId,
          })
          const result = cb(context, currentValue)

          if (result instanceof Promise) {
            this.actionChain.emit('operator:async', {
              type,
              name,
              path,
              isAsync: true,
              ...context.__execution,
              operatorId,
            })
            return result.then((promiseResult) => {
              this.actionChain.emit('operator:end', {
                type,
                name,
                path,
                ...context.__execution,
                isAsync: true,
                result: promiseResult,
                operatorId,
              })
              return promiseResult
            })
          }

          this.actionChain.emit('operator:end', {
            type,
            name,
            path,
            ...context.__execution,
            isAsync: false,
            result: result,
            operatorId,
          })

          return result
        }

        if (prevResult instanceof Promise) {
          return prevResult.then(produceResult)
        }

        return produceResult(prevResult)
      },
    ]
  }
}
