// overmind-devtools-client/src/components/Chart/styles.ts
import { css } from 'emotion'

import { colors } from '../../theme'

export const wrapper = css({
  position: 'relative',
  display: 'table',
  margin: '2rem',
  borderSpacing: '0 0.25rem',
})

export const stateItem = css({
  display: 'table-row',
})

export const stateNameCell = css({
  display: 'table-cell',
  padding: '0.5rem 0 ',
})

export const stateLineCell = css({
  position: 'relative',
  display: 'table-cell',
  width: '1rem',
})

export const stateName = css({
  borderRadius: 3,
  padding: '0.5rem',
  boxSizing: 'border-box',
  width: '100%',
  textAlign: 'center',
  backgroundColor: colors.foreground,
})

export const selectedLine = css({
  position: 'absolute',
  height: 2,
  backgroundColor: colors.foreground,
  width: '100%',
  top: '2rem',
})

export const entryCell = css({
  position: 'relative',
  display: 'table-cell',
  width: '1rem',
})

export const onName = css({
  borderRadius: 3,
  padding: '0.5rem',
  margin: '0.25rem 0',
  boxSizing: 'border-box',
  width: '100%',
  textAlign: 'center',
  backgroundColor: colors.foreground,
  opacity: 0.5,
})

export const activeAction = css({
  opacity: 1,
})

export const onCell = css({
  display: 'table-cell',
  padding: '0.5rem ',
  border: '2px solid ' + colors.foreground,
  borderRadius: 3,
})
