// overmind-devtools/src/styles/text.ts
import { css } from 'emotion'

export const primary = css`
  color: var(--color-primary);
`

export const blue = css`
  color: var(--color-blue);
`

export const yellow = css`
  color: var(--color-yellow);
`

export const purple = css`
  color: var(--color-purple);
`
export const white = css`
  color: var(--color-white-1);
`

export const white2 = css`
  color: var(--color-white-2);
`

export const black = css`
  color: var(--color-black-1);
`

export const monospace = css`
  font-family: monospace;
  line-height: 16px;
`

export const header = css`
  font-size: var(--font-size-6);
  text-transform: none;
  font-weight: bold;
  padding: var(--padding-5) 0;
`

export const denseHeader = css`
  ${header}
  padding: var(--padding-3) 0;
`

export const title = css`
  font-size: var(--font-size-4);
  text-transform: none;
  font-weight: bold;
  padding: var(--padding-4) 0;
`

export const denseTitle = css`
  ${title}
  padding: var(--padding-2) 0;
`

export const label = css`
  font-size: var(--font-size-3);
  text-transform: uppercase;
  font-weight: normal;
  padding: var(--padding-3) 0;
`

export const denseLabel = css`
  ${label}
  padding: var(--padding-2) 0;
`

export const normal = css`
  font-size: var(--font-size-3);
  text-transform: none;
  font-weight: normal;
  padding: var(--padding-3) 0;
`

export const denseNormal = css`
  ${normal}
  padding: var(--padding-2) 0;
`

export const description = css`
  font-size: var(--font-size-2);
  text-transform: none;
  font-weight: normal;
  padding: none;
`

export const hint = css`
  font-size: var(--font-size-1);
  text-transform: none;
  font-weight: normal;
  padding: none;
`
