// overmind-devtools/src/components/Flushes/elements.tsx
import styled from '../../styled-components'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  padding: ${({ theme }) => theme.padding.normal};
  height: 100%;
  box-sizing: border-box;
  overflow-y: scroll;
`

export const Panels = styled.div`
  display: flex;
  justify-content: space-around;
`

export const Panel = styled.div`
  padding: ${({ theme }) => theme.padding.smaller};
`

export const FlushWrapper = styled.div`
  margin: ${({ theme }) => `0 ${theme.padding.normal}`};
  margin-bottom: ${({ theme }) => theme.padding.smaller};
  border-top: 1px solid
    ${({ theme }) => theme.color.fade(theme.color.black, 0.9)};
`

export const Mutation = styled.div`
  display: flex;
  border: 1px solid ${({ theme }) => theme.color.fade(theme.color.black, 0.9)};
  border-top: 0;
  padding: ${({ theme }) => `0 ${theme.padding.smaller}`};
  align-items: center;
  > * {
    margin-right: 10px;
  }
  > *:nth-child(1) {
    flex: 0 0 150px;
  }
  > *:nth-child(2) {
    flex: 1;
  }
  > *:nth-child(3) {
    flex: 0 0 250px;
  }
`

export const Method = styled.span`
  color: ${({ theme }) => theme.color.primary};
  font-weight: bold;
`
