// overmind-devtools-client/src/components/SplitPane/index.tsx
import { Child, createElement } from 'overmind-components'

import { Component } from '../../overmind'
import * as styles from './styles'

type Props = {
  minSize: number
  defaultSize: number
  onChange: (size: number) => void
  children: [Child, Child]
}

type State = {
  isDragging: boolean
  size: number
  initialX: number
}

const SplitPane: Component<Props, State> = (
  { children, minSize, defaultSize, onChange },
  { state, changeState }
) => {
  return (
    <self
      onMount={() => {
        function onMouseMove(event) {
          if (state.isDragging) {
            const newSize = defaultSize + (event.clientX - state.initialX)

            changeState({
              size: Math.max(minSize, newSize),
            })
          }
        }
        function onMouseUp() {
          changeState({
            isDragging: false,
          })
          onChange(state.size)
        }

        window.addEventListener('mousemove', onMouseMove)
        window.addEventListener('mouseup', onMouseUp)

        return () => {
          window.removeEventListener('mousemove', onMouseMove)
          window.removeEventListener('mouseup', onMouseUp)
        }
      }}
      className={styles.wrapper}
    >
      <div
        style={{
          width: state.size + 'px',
        }}
      >
        {children[0]}
      </div>
      <div
        className={styles.splitter}
        onMouseDown={(event) => {
          changeState({
            isDragging: true,
            initialX: event.clientX,
          })
        }}
      />
      {children[1]}
    </self>
  )
}

SplitPane.getInitialState = ({ defaultSize }) => ({
  size: defaultSize,
  isDragging: false,
  initialX: 0,
})

export default SplitPane
