// overmind-devtools-client/webpack.config.js
const path = require('path')

module.exports = {
  devtool: 'inline-source-map',
  mode: 'development',
  entry: path.resolve('src', 'index.tsx'),
  output: {
    path: path.resolve('dist'),
    filename: 'bundle.js',
  },
  devServer: {
    disableHostCheck: true,
  },
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        loader: 'babel-loader',
        include: path.resolve('src'),
        options: {
          presets: [
            [
              '@babel/preset-typescript',
              {
                isTSX: true,
                jsxPragma: 'createElement',
                allExtensions: true,
              },
            ],
            [
              '@babel/preset-react',
              {
                pragma: 'createElement',
              },
            ],
            '@babel/preset-env',
          ],
          plugins: [
            '@babel/plugin-proposal-class-properties',
            '@babel/plugin-transform-runtime',
          ],
        },
      },
      {
        test: /\.(woff2)$/,
        use: [
          {
            loader: 'url-loader',
          },
        ],
      },
    ],
  },
  resolve: {
    extensions: ['.js', '.tsx', '.ts'],
  },
  plugins: [],
}
