// overmind-devtools/src/components/Apps/styles.ts
import { css } from 'emotion'

export const wrapper = css`
  position: relative;
  display: flex;
  background-color: var(--color-white-1);
  height: 50px;
  justify-content: center;
  align-items: center;
`

export const app = css`
  outline: none;
  border: 0;
  width: 100%;
  cursor: pointer;
  background-color: var(--color-white-1);
  display: flex;
  align-items: center;
  color: var(--color-black-1);
  padding: var(--padding-3) var(--padding-4);
  white-space: nowrap;
`

export const appSelected = css`
  background-color: var(--color-white-2);
`

export const currentApp = (color: string) => css`
  border: 2px solid ${color};
  border-radius: 50%;
  background-color: #fff;
  width: 30px;
  height: 30px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  font-size: var(--font-size-2);
  font-weight: bold;
`

export const overlay = css`
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  z-index: 99999;
`

export const appsList = css`
  position: absolute;
  left: calc(100% + 1px);
  top: 0;
  background-color: #fff;
  border-right: 1px solid var(--color-white-2);
  border-bottom: 1px solid var(--color-white-2);
`
