// overmind-devtools/src/client/app/mutations.ts
import { Apps, State, Message, Tab } from './state'

export const setApps = (apps: Apps, state: State) => (state.apps = apps || {})

export const setCurrentPort = (currentPort: string, state: State) => {
  if (currentPort) {
    state.currentPort = currentPort
  } else if (Object.keys(state.apps).length) {
    state.currentPort = Object.keys(state.apps)[0]
  }
}

export const setError = (error: string, state: State) => (state.error = error)

export const setAppLoaded = (_, state: State) => (state.isLoading = false)

export const setNewPortValue = (value: string, state: State) =>
  (state.newPortValue = value)

export const addNewApp = (_, state: State) =>
  (state.apps[state.newPortValue] = {
    name: null,
    port: state.newPortValue,
    messages: [],
  })

export const resetNewPortValue = (_, state: State) => (state.newPortValue = '')

export const addMessageFromClient = (message: Message, state: State) => {
  state.apps[message.port].messages = state.apps[message.port].messages.concat(
    message.message
  )
}

export const changeTab = (tab: Tab, state: State) => (state.currentTab = tab)
