// overmind-devtools/src/client/app/providers.ts
import * as jsonStorage from 'electron-json-storage'
import BackendConnector from '../BackendConnector'

const backendConnector = new BackendConnector()

export const connector: {
  addPort: typeof backendConnector.addPort
  removePort: typeof backendConnector.removePort
  sendMessage: typeof backendConnector.sendMessage
  relaunch: typeof backendConnector.relaunch
} = {
  addPort: backendConnector.addPort.bind(backendConnector),
  removePort: backendConnector.removePort.bind(backendConnector),
  sendMessage: backendConnector.sendMessage.bind(backendConnector),
  relaunch: backendConnector.relaunch.bind(backendConnector),
}

export const storage = {
  clear() {
    return new Promise((resolve, reject) => {
      jsonStorage.clear((error) => {
        if (error) {
          reject(error)
        } else {
          resolve()
        }
      })
    })
  },
  get(key: string) {
    return new Promise((resolve, reject) => {
      jsonStorage.get(key, (err, data) => {
        if (err) {
          reject(err)
        } else {
          resolve(
            // Insanely weird that when "no value", an empty
            // object is returned. Should be null
            typeof data === 'object' &&
            !Array.isArray(data) &&
            data !== null &&
            Object.keys(data).length === 0
              ? null
              : data
          )
        }
      })
    })
  },
  set(key, data) {
    return new Promise((resolve, reject) => {
      jsonStorage.set(key, data, (err) => {
        if (err) {
          reject(err)
        } else {
          resolve()
        }
      })
    })
  },
}
