// proxy-state-tree/src/types.ts
export interface IProxifier<T extends object> {
  proxify(state: T, path: string): T
  trackPath(path: string): void
}

export interface IMutation {
  method: string
  path: string
  args: any[]
}

export interface IMutationCallback {
  (mutation: IMutation, paths: Set<string>, flushId: number): void
}

export interface ITrackMutationTree<T extends object> {
  addMutation(mutation: IMutation, objectChangePath?: string): void
  onMutation(callback: IMutationCallback): void
  canTrack(): boolean
  canMutate(): boolean
  dispose(): ITrackMutationTree<T>
  master: IProxyStateTree<T>
  proxifier: IProxifier<T>
  state: T
}

export interface ITrackCallback {
  (
    mutations: IMutation[],
    paths: Set<string>,
    flushId: number,
    isAsync: boolean
  ): void
}

export interface ITrackScopedCallback<T extends object> {
  (tree: ITrackStateTree<T>): any
}

export interface ITrackStateTree<T extends object> {
  addTrackingPath(path: string): void
  track(cb?: ITrackCallback): ITrackStateTree<T>
  trackScope(scope: ITrackScopedCallback<T>, callback?: ITrackCallback): any
  canTrack(): boolean
  canMutate(): boolean
  dispose(): ITrackStateTree<T>
  master: IProxyStateTree<T>
  proxifier: IProxifier<T>
  state: T
  pathDependencies: Set<string>
  callback: ITrackCallback
}

export interface IOptions {
  devmode?: boolean
  dynamicWrapper?: Function
}

export interface IFlushCallback {
  (
    mutations: IMutation[],
    paths: string[],
    flushId: number,
    isAsync: boolean
  ): void
}

export type TTree = ITrackMutationTree<any> | ITrackStateTree<any>

export interface IProxyStateTree<T extends object> {
  addMutation(mutation: IMutation, objectChangePath?: string): void
  addPathDependency(path: string, callback: ITrackCallback): void
  removePathDependency(path: string, callback: ITrackCallback): void
  getTrackStateTree(): ITrackStateTree<T>
  getMutationTree(): ITrackMutationTree<T>
  changeTrackStateTree(tree: ITrackStateTree<T>): void
  disposeTree(proxy: TTree): void
  onMutation(cb: IMutationCallback): void
  onFlush(cb: IFlushCallback): void
  rescope(value: any, tree: TTree): any
  sourceState: T
  state: T
  options: IOptions
  pathDependencies: {
    [path: string]: Set<ITrackCallback>
  }
  objectChanges: Set<string>
  master: IProxyStateTree<T>
  proxifier: IProxifier<T>
  currentTree: TTree
  previousTree: TTree
  mutationTree: ITrackMutationTree<T>
  currentFlushId: number
}
