// overmind-devtools/src/components/common/Table/index.tsx
import { h } from 'overmind-components'
import * as styles from './styles'
import { Component } from '../../../app'

type Header = {
  title: string
  width: string
}

type TableProps = {
  headers?: Header[]
}

const Table: Component<TableProps> = ({ headers, children }) => (
  <table className={styles.tableComponent}>
    {headers ? (
      <thead>
        <Row>
          {headers.map((header, index) => (
            <th
              className={styles.header}
              key={header.title + index}
              style={{ width: header.width }}
            >
              {header.title}
            </th>
          ))}
        </Row>
      </thead>
    ) : null}
    <tbody>{children}</tbody>
  </table>
)

export const Row: Component = ({ children }) => <tr>{children}</tr>

type CellProps = {
  wordwrap?: string
  mono?: boolean
  colSpan?: number
}

export const Cell: Component<CellProps> = ({
  wordwrap,
  mono,
  colSpan,
  children,
}) => (
  <td className={styles.cellComponent(mono, wordwrap)} colSpan={colSpan}>
    {children}
  </td>
)

export default Table
