// overmind-devtools/src/components/common/Table/index.tsx
import * as React from 'react'
import {
  TableComponent,
  Headers,
  Header,
  Body,
  RowComponent,
  CellComponent,
} from './elements'

type Header = {
  title: string
  width: string
}

type TableProps = {
  headers?: Header[]
}

const Table: React.SFC<TableProps> = ({ headers, children }) => (
  <TableComponent>
    {headers ? (
      <Headers>
        <Row>
          {headers.map((header, index) => (
            <Header key={header.title + index} width={header.width}>
              {header.title}
            </Header>
          ))}
        </Row>
      </Headers>
    ) : null}
    <Body>{children}</Body>
  </TableComponent>
)

export const Row: React.SFC = ({ children }) => (
  <RowComponent>{children}</RowComponent>
)

type CellProps = {
  wordwrap?: string
  mono?: boolean
  colSpan?: number
}

export const Cell: React.SFC<CellProps> = ({
  wordwrap,
  mono,
  colSpan,
  children,
}) => (
  <CellComponent wordwrap={wordwrap} mono={mono} colSpan={colSpan}>
    {children}
  </CellComponent>
)

export default Table
