// overmind-devtools/src/components/ActionPath/elements.tsx
import styled from '../../styled-components'
import { Colors } from 'overmind-themes'

export const Path = styled<
  {
    visible?: boolean
  },
  'div'
>('div')`
  position: relative;
  height: ${({ visible }) => (visible ? '100%' : '0')}
  visibility: ${({ visible }) => (visible === false ? 'hidden' : 'visible')};
`

export const PathLine = styled<
  {
    borderColor?: keyof Colors
  },
  'div'
>('div')`
  position: absolute;
  top: 50%;
  left: 0;
  z-index: 0;
  width: 100%;
  border-bottom: 1px solid
    ${({ theme, borderColor }) =>
      borderColor ? theme.color[borderColor] : theme.color.black};
`

export const PathText = styled<
  {
    borderColor?: keyof Colors
    visible?: boolean
  },
  'div'
>('div')`
  position: relative;
  z-index: 1;
  margin: ${({ theme }) => `0 ${theme.padding.smaller} 0 0`};
  padding: ${({ theme }) =>
    `${theme.padding.smallest} ${theme.padding.smaller}`};
  border: 1px solid ${({ theme, borderColor }) => theme.color[borderColor]};
  background-color: #fff;
  border-radius: ${({ theme }) => theme.borderRadius.normal};
  visibility: ${({ visible }) => (visible ? 'visible' : 'hidden')};
`
