// overmind/src/index.ts
import { ActionChain } from 'action-chain'
import ProxyStateTree from 'proxy-state-tree'
import Devtools, { Message, safeValue } from './Devtools'
import Action, { IValueAction, INoValueAction } from './Action'
export { default as compose } from './compose'
export { default as derived } from './derived'
export { default as computed } from './computed'

type Configuration<State, Providers, Actions> = {
  state?: State
  providers?: Providers
  actions?: Actions
}

type Options = {
  devtools?: string
}

export type ActionsCallback<Providers, State> = (
  action: IAction<State, Providers & { state: State }>
) => any

export type TContext<State, Providers = {}> = Providers & {
  state: State
}

export interface IAction<State, Context> {
  <InitialValue = void>(): [InitialValue] extends [void]
    ? INoValueAction<State, Context, InitialValue>
    : IValueAction<State, Context, InitialValue>
}

export default class App<
  State extends object,
  Providers extends object,
  Actions extends
    | {
        [namespace: string]: ActionsCallback<Providers, State>
      }
    | ActionsCallback<Providers, State>
> {
  private proxyStateTree: ProxyStateTree
  devtools: Devtools
  actions: Actions extends {
    [namespace: string]: ActionsCallback<Providers, State>
  }
    ? { [Namespace in keyof Actions]: ReturnType<Actions[Namespace]> }
    : Actions extends ActionsCallback<Providers, State>
      ? ReturnType<Actions>
      : any
  state: State
  constructor(
    configuration: Configuration<State, Providers, Actions>,
    options: Options = {}
  ) {
    const proxyStateTree = new ProxyStateTree(configuration.state || {}, {
      dynamicWrapper: (proxyStateTree, path, func) =>
        func(actionChain, proxyStateTree, path),
    })
    const actionChain = new ActionChain(
      Object.assign(
        {
          state: proxyStateTree.get(),
        },
        configuration.providers || ({} as Providers)
      ),
      {
        providerExceptions: ['state'],
      }
    )
    const action = function<InitialValue>(): [InitialValue] extends [void]
      ? INoValueAction<State, Providers & { state: State }, InitialValue>
      : IValueAction<State, Providers & { state: State }, InitialValue> {
      return new Action(proxyStateTree, actionChain) as any
    }

    actionChain.on('operator:async', () => {
      if (this.devtools) {
        this.devtools.send({
          type: 'flush',
          data: { mutations: proxyStateTree.mutations },
        })
      }
      proxyStateTree.flush()
    })
    actionChain.on('action:end', () => {
      if (this.devtools) {
        this.devtools.send({
          type: 'flush',
          data: { mutations: proxyStateTree.mutations },
        })
      }
      proxyStateTree.flush()
    })

    this.state = proxyStateTree.get()
    this.actions =
      typeof configuration.actions === 'function'
        ? (configuration.actions as ActionsCallback<Providers, State>)(
            action as IAction<State, Providers & { state: State }>
          )
        : (Object.keys(configuration.actions || {}).reduce(
            (aggr, namespace) =>
              Object.assign(aggr, {
                [namespace]: configuration.actions[namespace](action as IAction<
                  State,
                  Providers & { state: State }
                >),
              }),
            {}
          ) as any)
    this.proxyStateTree = proxyStateTree

    if (options.devtools && typeof window !== 'undefined') {
      this.initializeDevtools(options.devtools, actionChain)
    }
  }
  private initializeDevtools(host, actionChain) {
    const devtools = new Devtools()
    devtools.connect(
      host,
      (message: Message) => {
        // To use for communication from devtools app
      }
    )
    devtools.send({
      type: 'init',
      data: {
        state: this.proxyStateTree.get(),
      },
    })
    actionChain.on('action:start', (data) =>
      devtools.send({
        type: 'action:start',
        data,
      })
    )
    actionChain.on('operator:start', (data) =>
      devtools.send({
        type: 'operator:start',
        data,
      })
    )
    actionChain.on('operator:end', (data) =>
      devtools.send({
        type: 'operator:end',
        data: {
          ...data,
          result: safeValue(data.result),
        },
      })
    )
    actionChain.on('action:end', (data) =>
      devtools.send({
        type: 'action:end',
        data,
      })
    )
    actionChain.on('provider', (data) =>
      devtools.send({
        type: 'provider',
        data,
      })
    )
    actionChain.on('mutations', (data) =>
      devtools.send({
        type: 'mutations',
        data,
      })
    )
    actionChain.on('derived', (data) =>
      devtools.send({
        type: 'derived',
        data,
      })
    )
    actionChain.on('computed', (data) =>
      devtools.send({
        type: 'computed',
        data,
      })
    )
    this.devtools = devtools
  }
  trackState() {
    return this.proxyStateTree.startPathsTracking()
  }
  clearTrackState(id: number) {
    return this.proxyStateTree.clearPathsTracking(id)
  }
  addMutationListener(paths, cb) {
    return this.proxyStateTree.addMutationListener(paths, cb)
  }
}
