// overmind/src/types.ts
import { ResolveActions, ResolveState, TBaseContext } from './internalTypes'
import { Overmind } from './'

/** ===== PUBLIC API
 */
export { EventType } from './internalTypes'

export type Configuration = {
  onInitialize?: any
  state?: {}
  effects?: {}
  actions?: {}
}

export type BaseApp = {
  state: {}
  effects: {}
  actions: {}
}

export type TStateObject =
  | {
      [key: string]:
        | TStateObject
        | string
        | IDerive<any, any, any>
        | number
        | boolean
        | object
        | null
        | undefined
    }
  | undefined

export interface IConfig<ThisConfig extends Configuration> {
  state: ThisConfig['state'] & {}
  actions: ThisConfig['actions'] & {}
  effects: ThisConfig['effects'] & {}
}

// This is the type of the `app` argument passed in components.
export type TApp<ThisConfig extends Configuration> = {
  // Resolves `Derive` types in state.
  state: ResolveState<ThisConfig['state']>
  actions: ResolveActions<ThisConfig['actions']>
  effects: ThisConfig['effects']
}

// This is the value passed through operators
export type TOperatorContext<
  ThisConfig extends Configuration,
  Value
> = TBaseContext<ThisConfig> & {
  value: Value
}

// This is the first argument of an action
export type TContext<ThisConfig extends Configuration> = TBaseContext<
  ThisConfig
>

export interface IAction<ThisConfig extends Configuration, Value> {
  (context: TContext<ThisConfig>, value: Value): any
}

export type IOperator<
  ThisConfig extends Configuration,
  Input,
  Output = Input
> = (
  err: Error | null,
  val: TOperatorContext<ThisConfig, Input>,
  next: (err: Error | null, val?: TOperatorContext<ThisConfig, Output>) => void,
  final?: (err, Error, val?: TOperatorContext<ThisConfig, Output>) => void
) => void

export type IDerive<
  ThisConfig extends Configuration,
  Parent extends TStateObject,
  Value
> = (
  parent: ResolveState<Parent>,
  state: ResolveState<ThisConfig['state'] & {}>
) => Value

export interface IOnInitialize<ThisConfig extends Configuration> {
  (context: TContext<ThisConfig>, value: Overmind<ThisConfig>): void
}
