// overmind/src/types.ts
import { ResolveActions, ResolveState, TBaseContext } from './internalTypes'

/** ===== PUBLIC API
 */
export { EventType } from './internalTypes'

export type Configuration = {
  onInitialize?: any
  state?: {}
  effects?: {}
  actions?: {}
  reactions?: {}
}

export type BaseApp = {
  state: {}
  effects: {}
  actions: {}
}

export interface TApp<Config extends Configuration> {
  // Resolves `Derive` types in state.
  state: ResolveState<Config['state'] & {}>
  // Transform actions into callable functions.
  actions: ResolveActions<Config['actions']>
  effects: Config['effects'] & {}
}

export type TValueContext<App extends BaseApp, Value> = TBaseContext<App> & {
  value: Value
}

export type TAction<App extends BaseApp, Value> = (
  context: TValueContext<App, Value>
) => any

export type TOperator<App extends BaseApp, Input, Output> = (
  err: Error | null,
  val: TValueContext<App, Input>,
  next: (err: Error | null, val?: TValueContext<App, Output>) => void,
  final?: (err, Error, val?: TValueContext<App, Output>) => void
) => void

export type TDerive<App extends BaseApp, Parent extends object, Value> = (
  parent: Parent,
  state: App['state']
) => Value

export type TReaction<App extends BaseApp> = (
  reaction: (
    getState: (state: App['state']) => any,
    action: TAction<App, void> | TOperator<App, void, any>
  ) => any
) => any
