// overmind-devtools/src/components/ActionOperator/styles.ts
import { css, keyframes } from 'emotion'

export const operator = (borderColor: string) => css`
  flex: 1;
  display: flex;
  flex-direction: column;
  padding: 0 var(--padding-4);
  cursor: default;
  border: 1px solid var(--color-dark-1);
  background-color: var(--color-dark-2);
  border-radius: var(--border-radius-1);
`

export const operatorClickable = css`
  cursor: pointer;
`

export const operatorType = css`
  color: var(--color-white-4);
`

export const chevron = css`
  color: var(--color-white-4);
`

export const notVisible = css`
  visibility: hidden;
`

export const operatorHeader = css`
  display: flex;
  align-items: center;
  > *:first-child {
    margin-right: 10px;
  }
  > *:nth-child(3) {
    margin-left: auto;
    color: var(--color-white-2);
  }
`

export const operatorItem = css`
  display: flex;
  align-items: center;
  margin-bottom: var(--padding-3);
  > * {
    margin-right: 10px;
  }
`

export const effectError = css`
  background-color: var(--color-red);
  color: var(--color-white-1);
  padding: var(--padding-1) var(--padding-3);
  font-size: var(--font-size-2);
  border-radius: 2px;
`

const pulse = keyframes`
from {
  opacity: 0.5;
}
to {
  opacity: 1;
}
`

export const pendingOperatorItem = css`
  animation: ${pulse} 1s ease-out infinite;
  animation-direction: alternate;
`
