// overmind-devtools/src/components/common/Text/elements.tsx
import styled from '../../../styled-components'

export type Variant = 'header' | 'title' | 'label' | 'text'

const styles = {
  header: {
    fontSize: 'largest',
    textTransform: 'none',
    fontWeight: 'bold',
    padding: 'large',
    densePadding: 'small',
  },
  title: {
    fontSize: 'large',
    textTransform: 'none',
    fontWeight: 'bold',
    padding: 'normal',
    densePadding: 'smaller',
  },
  label: {
    fontSize: 'medium',
    textTransform: 'uppercase',
    fontWeight: 'normal',
    padding: 'small',
    densePadding: 'smaller',
  },
  text: {
    fontSize: 'normal',
    textTransform: 'none',
    fontWeight: 'normal',
    padding: 'small',
    densePadding: 'smaller',
  },
}

export const TextElement = styled<
  {
    variant?: Variant
    dark?: boolean
    dense?: boolean
  },
  'p'
>('p')`
  margin: 0;
  text-transform: ${({ theme, variant }) => styles[variant].textTransform};
  padding: ${({ theme, variant, dense }) =>
    dense
      ? `${theme.padding[styles[variant].densePadding]} 0`
      : `${theme.padding[styles[variant].padding]} 0`};
  color: ${({ theme, dark }) => (dark ? theme.color.white : theme.color.black)};
  font-size: ${({ theme, variant }) =>
    theme.fontSize[styles[variant].fontSize]};
  font-weight: ${({ theme, variant }) => styles[variant].fontWeight};
`
