// overmind-devtools/src/app/actions.ts
import { Action } from 'overmind'
import * as mutations from './mutations'
import * as operations from './operations'
import { Message, Tab } from './types'

const onMessage: Action<Message> = (action) =>
  action()
    .mutate(mutations.performMutationsByMessageType)
    .mutate(mutations.addMessagesFromClient)

export const loadDevtools: Action = (action) =>
  action()
    // .do(({ storage }) => storage.clear())
    // Do a check if the current app matches the keys of the ???
    .map(operations.getAppsFromStorage)
    .mutate(mutations.setApps)
    .map(operations.getCurrentPortFromStorage)
    .mutate(mutations.setCurrentPort)
    .mutate(mutations.setAppLoaded)
    .do(operations.connectCurrentPort(onMessage(action as any)))

export const setError: Action<string> = (action) =>
  action().mutate(mutations.setError)

export const changeNewPortValue: Action<string> = (action) =>
  action()
    .map(operations.toNumber)
    .mutate(mutations.setNewPortValue)

export const addPort: Action = (action) =>
  action()
    .map(operations.getNewPortFromState)
    .mutate(mutations.setCurrentPort)
    .mutate(mutations.addNewApp)
    .mutate(mutations.resetNewPortValue)
    .do(operations.storeApps)
    .do(operations.connectCurrentPort(onMessage))

export const changeTab: Action<Tab> = (action) =>
  action().mutate(mutations.changeTab)

export const toggleExpandState: Action<string[]> = (action) =>
  action().mutate(mutations.toggleExpandStatePath)

export const selectAction: Action<string> = (action) =>
  action()
    .mutate(mutations.toggleActionItemCollapse)
    .mutate(mutations.selectAction)

type Collapsed = {
  isCollapsed: boolean
}

export const toggleCollapsed: Action<Collapsed> = (action) =>
  action()
    .filter(operations.isNotExpandingAllActions)
    .mutate(mutations.toggleCollapsed)

export const configurePort: Action = (action) =>
  action().mutate(mutations.configurePort)

export const cancelConfigurePort: Action = (action) =>
  action().mutate(mutations.cancelConfigurePort)

export const removeApp: Action = (action) =>
  action()
    .filter(operations.confirm('Are you sure you want to remove the app?'))
    .do(operations.removeCurrentPort)
    .mutate(mutations.removeApp)
    .do(operations.storeApps)
    .do(operations.connectCurrentPort(onMessage(action as any)))

export const selectPort: Action<string> = (action) =>
  action()
    .mutate(mutations.selectPort)
    .do(operations.connectCurrentPort(onMessage(action as any)))

export const toggleExpandAllActions: Action = (action) =>
  action().mutate(mutations.toggleExpandAllActions)
