// overmind-devtools/src/app/derived.ts
import { State, Component } from './state'

export const currentApp = (state: State) => state.apps[state.currentPort]

export const componentsMounted = (state: State) =>
  Object.keys(state.currentApp.components).reduce(
    (aggr, key) => {
      if (state.currentApp.components[key].isMounted) {
        return aggr.concat(state.currentApp.components[key])
      }

      return aggr
    },
    [] as Component[]
  )

export const componentsUpdateCount = (state: State) =>
  state.componentsMounted.reduce(
    (aggr, component) => aggr + component.updateCount,
    0
  )

export const componentsStatePathCount = (state: State) =>
  state.componentsMounted.reduce(
    (aggr, component) => aggr + component.paths.length,
    0
  )

export const flushes = (state: State) =>
  Object.keys(state.currentApp.flushes)
    .sort(
      (idA, idB) =>
        state.currentApp.flushes[idB].flushId -
        state.currentApp.flushes[idA].flushId
    )
    .map((id) => state.currentApp.flushes[id])

export const flushesMutationsCount = (state: State) =>
  state.flushes.reduce((aggr, flush) => aggr + flush.mutations.length, 0)

export const flushesStatePathCount = (state: State) =>
  state.flushes.reduce((aggr, flush) => {
    return flush.mutations.reduce(
      (aggr, mutation) =>
        aggr.includes(mutation.path) ? aggr : aggr.concat(mutation.path),
      aggr
    )
  }, []).length

export const currentAction = (state: State) =>
  state.currentApp.actions[state.currentApp.currentActionId]
