// overmind-devtools/src/app/index.ts
import Overmind, {
  TAction,
  TApp,
  TContext,
  TDerive,
  TMutate,
  TReaction,
} from 'overmind'
import createConnect, { TConnect } from 'overmind-react'

import * as actions from './actions'
import * as effects from './effects'
import * as state from './state'

const config = {
  effects,
  actions,
  state,
}

type App = TApp<typeof config>

export type Action<Value = void, ReturnValue = any> = TAction<
  App,
  Value,
  ReturnValue
>

export type Mutate<Value = any> = TMutate<App, Value>

export type Context<Value> = TContext<App, Value>

export type Map<Value, ReturnValue = Value> = (
  ctx: Context<Value>
) => ReturnValue

export type Filter<Value = any> = (ctx: Context<Value>) => boolean

export type When<Value = any> = (ctx: Context<Value>) => boolean

export type Run<Value = any> = (ctx: Context<Value>) => void

export type Fork<Value = any> = (ctx: Context<Value>) => string

export type Attempt<Value, ReturnValue> = (ctx: Context<Value>) => ReturnValue

export type Derive<Value> = TDerive<App, Value>

export type Reaction = TReaction<App>

const app = new Overmind(config, {
  devtools: false,
})

export type Connect = TConnect<typeof app>

export const connect = createConnect(app)

export default app
